### ISSUE TITLE
Windows 3.1 absolute mouse

### ISSUE THREAD

**Issue by stuaxo** [dosemu2/dosemu2#1552]
https://github.com/NattyNarwhal/vmwmouse

There is a new mouse driver for Windows 3.1 in VMware, it could provide a basis for more seamless mouse support in dosemu2 in Windows 3.1.

Apologies if this isn't needed, I'm not in front of a computer to test right now.

**Comment by stuaxo** [comment]
Reading the doc linked from there about the VMware back door
https://wiki.osdev.org/VMware_tools#Absolute_Mouse_Coordinates

The API seems to have some things that are useful.
I realise dosemu has its own mechanism via its special interrupt, but it's interesting that according to the doc Qemu implements some of the functionality.

I guess this is as close to a sort of standard it's possible to get in host/guest communication.

**Comment by stsp** [comment]
So we can add a new mouse
back-end to dosemu2 for better
mouse support under vmware
and qemu. May be good for
fdpp-buildroot project, but
fdpp-buildroot project is not
yet showing its value, other than
to debug KVM problems.

Or we can add similar mechanism
for guest-side mouse tracking,
but to only support this win3.1
mouse driver, whereas we already
have our own which is not any
worse.

Not sure which of the above 2
possibilities do you have in mind.
Both tasks are quite simple and
practically almost useless, but
the patches are welcome. Why
not to support something new.

**Comment by stuaxo** [comment]
Thanks for more fully thinking this through.

One practical thing I will do is some tests to verify our driver works as this does in the gif in their README, when it comes to moving the mouse in and out of the DOSEMU2 Window, I seem to recall various instances where it's not as well synced, but that could be due to my environment not DOSEMU2 itself.

**Comment by stsp** [comment]
https://github.com/dosemu2/win31-mouse-driver/tree/master/out

**Comment by stsp** [comment]
OK, there seem to be no qemu
support for that mouse option.
It seems to be specific to vmware.
qemu seems to only support
USB tablet for this, but under
fdpp-buildroot that usb-tablet
sucks (mouse becomes way
too sensitive to be usable).
I wonder if it is the qemu bug
or something else.

**Comment by stsp** [comment]
Should now be fixed.
I added vmware and qemu mouses
support to fdpp-buildroot, and
submitted the patch to sdl with
qemu mouse/tablet support.
This is probably the best we
can aim for.
Given that this vmware mouse
is not supported by qemu, I see
zero need doing anything else.

**Comment by stuaxo** [comment]
Wow you added this before I managed to reply :)

I always thought tablet mode would = absolute mouse, but maybe not ?

I probably need to setup some Windows 3.1 VMs and see how the mouse works in them, which may result in a feature request to Qemu, but if tablet mode works could be moot.

Cheers for implementing this I'll do some testing.

**Comment by stsp** [comment]
Doesn't work for me.

**Comment by NattyNarwhal** [comment]
Author of the original repo here, I'm going to try this in QEMU soon. Ping me if you have any questions about my source/the interface. I've been reading the [QEMU impl source](https://github.com/qemu/qemu/blob/master/hw/i386/vmmouse.c) too.

**Comment by stsp** [comment]
Yes, there is such impl.
But at least the following
should happen to consider
that working:
1. qemu should stop grabbing
mouse on the click
2. linux should detect the
mechanism and start reporting
absolute positions.

You can easily see both of the
above if you use `-usbdevice tablet`.
qemu stops grabbing mouse as
soon as the guest's tablet driver
initialized. With `-device vmmouse`
the grab is still happening, and
also I don't see in linux dmesg
that it detected vmmouse - it
seems to be still using relative
ps2 movements.
So maybe you can get that to
work properly, but tablet already
works (needs my sdl patch though).

**Comment by stsp** [comment]
The interface is also quite
strange:
```
Backdoor proc near
	
		mov eax, VMWARE_MAGIC
		mov dx, VMWARE_PORT
		in eax, dx
		ret
	
	Backdoor endp
```
Instead of having magic numbers
on port reads (never seen that before),
they (vmware) could just add the
new ps/2 commands for this.
Enter abs mode, leave abs mode,
get abs data, get abs resolution,
set abs resolution and all that.

In what resolution do they return
abs coords? Is it always similar
to guest screen resolution?

**Comment by stsp** [comment]
Already found an answer:
` EBX = x (0 - FFFFh scaled)`.

**Comment by stsp** [comment]
So when you do `in eax, dx`
they seem to alter all registers,
not just eax.
OK, I basically understand that
horrible interface now. :)

**Comment by stsp** [comment]
Well, dosemu2 is simply not
prepared for that kind of hacks.
All it does, is: `rAL = port_inb(a);`.
And if inside port_inb() we are
going to update all other registers,
then we need to decide which ones:
- real-mode
- prot mode
- the ones from CPU simulator
- the ones in JITted code

dosemu2 keeps all the above
cases separately. There is no
simple way to update all registers
when the asm command is not
asking for that.

**Comment by stsp** [comment]
Likewise, port handler doesn't
have any context for reaching
the "magic value". All context
that the port handler has or
should have, is the port number
to read and/or the value to write.
Getting into CPU context is
especially difficult under JITted
code. It is only available after
JIT exits. Of course the exits
are controlled by an ioperm
bitmask, and the one can make
them unconditional... And then
only look for rm or pm CPU
contexts...
But let's just say Hell No to
the ugly hacks. :)

**Comment by NattyNarwhal** [comment]
Yes, the backdoor does clobber EAX-EDX. I do notice the PS/2 stream is altered when you switch to absolute mode; I don't know if the protocol changes and it can communicate position that way or if the stream is inert and only used as an interrupt source; either way, it seems what you're supposed to do is check for PS/2 mouse events, then disregard their contents and do hypercalls instead (once to query, if successful, again to read). If you actually read a PS/2 packet (assuming it's in the normal format), it just looks like it's drifting to the bottom right, forever.

I suspect a cleaner interface can probably be done; probably as you say, extending the PS/2 interface. Obviously, that'd need driver work again.

**Comment by stsp** [comment]
> If you actually read a PS/2 packet (assuming it's in the normal format), it just looks like it's drifting to the bottom right, forever.

Maybe this is a trick to re-sync
the coords? We did that before
the abs coords appeared: you
drift to bottom-left for quite long,
and then, when you think the
cursor is in the corner, you move
down-right by the absolute coords.

> I suspect a cleaner interface can probably be done; probably as you say, extending the PS/2 interface.
> Obviously, that'd need driver work again.

I suppose unless you work
for vmware, its not possible
to implement a cleaner interface. :)
As first it have to be supported
by vmware...

**Comment by NattyNarwhal** [comment]
> Maybe this is a trick to re-sync the coords?

I've tried moving them for a while and it doesn't seem to be changing. (PS/2 button state does seem to work fine, so it makes me believe it's not being too misinterpreted.)

> I suppose unless you work for vmware, its not possible to implement a cleaner interface. :) As first it have to be supported by vmware...

I don't, but it's quite possible DOSEmu doesn't have to repeat their mistakes. :upside_down_face: 

Another note is that int 33h *can* actually deal with absolute coordinates, but in another way. I suspect you could possibly implement a `mouse.com` that deals in whatever absolute protocol and then teach the Windows mouse driver that 33h is relative.

**Comment by stsp** [comment]
But we already have that. :)
You can find our int33 win31 driver by
the link in that thread.

**Comment by stsp** [comment]
> implement a mouse.com

The way it is currently done,
requires no mouse.com because
dosemu is para-virtualizer, so
it just provides int33 on its own.
But to use our int33 driver under
qemu, indeed some patched
mouse.com would be needed.
But I won't advertice the vmware's
protocol to ctmouse authors
because of that register clobbering
thing, and I won't advertise the
usb tablet to them as it is quite
difficult to support under DOS...
So I suspect there currently is
no good option for mouse.com
(aka ctmouse.exe).

**Comment by jschwartzenberg** [comment]
> There is a new mouse driver for Windows 3.1 in VMware, it could provide a basis for more seamless mouse support in dosemu2 in Windows 3.1.

I'm curious what it could provide beyond the current `i33mouse` driver? Does it have scroll wheel support like what I demo here? https://schwart6.home.xs4all.nl/win31-scrollwheel.ogv
I don't think there are any other things that could be improved.

**Comment by NattyNarwhal** [comment]
Funny you mention that, because I'm implementing this in the VMware mouse driver now :grimacing:

**Comment by stsp** [comment]
How do we support wheel?
I can't find that neither in a
(very small) changelog, nor
in the code...
Magic.

**Comment by stsp** [comment]
Are you sure your demo shows
our i33mouse driver and not
something else?

**Comment by NattyNarwhal** [comment]
I can't speak for Julius' demo, but I've got my own implementation of this in a branch I haven't pushed yet, because it's kinda gross right now. What I basically do right now is:

* At least with the VMware interface, use the wheel data returned in `EDX `(8-bit signed int, -1, 0, or 1)
* Get the focused window (which is *not* ideal, because the focused window may not be the thing you want to scroll, and the thing you want to scroll may not be a focusable window. for another demo, I did a [`HWND_BROADCAST` to top-level windows](https://i.imgur.com/PIBLBTd.mp4) or `GetActiveWindow`, but those only work for again, top-level popups. I think I need to use `GetCursorPos` to get the position, then `WindowFromPoint` to get what's under the cursor, which incidentally would allow for )
* If not 0, do a `PostMessage` ([safe to call from an interrupt according to Raymond](https://devblogs.microsoft.com/oldnewthing/20211129-00/?p=105979)), sending a `WM_VSCROLL` to it (either `SB_LINEUP` or `SB_LINEDOWN`).

**Comment by stsp** [comment]
Would it be possible to make
this a separate driver or a dll,
and just provide another event_proc,
that our driver could call too,
as well as yours?

**Comment by NattyNarwhal** [comment]
Yes, it would have to be `GetProcAddress`'d, but I have to do that for `USER` anyways, because IIRC most drivers can only link against `KERNEL`.

**Comment by stsp** [comment]
What do you think about integrating
our (very small) int33 code into your
driver? That way it will work with both
vaware and, as a fall-back (when no
vmware API is there) - with mouse.com.
You definitely know how to code
for windows, and I do not.

**Comment by stsp** [comment]
I think @jschwartzenberg actually
didn't say the demo shows our
driver. Quite the opposite.
I misinterpreted.

**Comment by NattyNarwhal** [comment]
I actually ripped out the int 33h code (well, Microsoft's, not yours) from the driver to simplify it D: (since there was some runtime polymorphism stuff I wanted to reduce)

Maybe one way to make it shared w/o having to merge two drivers together would be something like:

1. There's a WHEEL.DLL or something (either a library or driver, whatever would work)
2. It exposes an API (i.e. a `wheel_proc` or whatever), takes a wheel direction
3. It does this dance with getting the target `HWND` and `PostMessage` for you
4. Our drivers can load these and if not present, skip calling the procedures

I have not tested this approach (first I want to get my implementation working before refactoring out into a separate DLL). Alternatively, the code could just simply be plopped in (since your 33h driver is also derived from the same source so I assume you're using MASM?) or adapted.

**Comment by stsp** [comment]
You rightly ripped int33 code as it
didn't support abs coords. Since your
driver is about abs coords, perhaps
it can only benefit from a secondary
abs coords source.
But if you prefer sharing a dll, that
might be fine too, as I'll be able to
copy/paste the integration code
from your driver.
Yeah, masm and same sdk.

**Comment by NattyNarwhal** [comment]
I'll have to experiment and see what works out. FWIW, there *is* still code to contact things on 33h (used primarily to tell the 33h driver to wake up when deinitializing), so I could shoehorn it into the ps2/vmware code path if there's a convenient API for reading i.e. wheel status and such.

**Comment by stsp** [comment]
Wheel is read from BH with the
event mask 80h iirc. So quite
convenient.

**Comment by jschwartzenberg** [comment]
> I think @jschwartzenberg actually didn't say the demo shows our driver. Quite the opposite. I misinterpreted.

I used a driver that handles serial mice with a scroll wheel together with my old Logitech Pilot Mouse+ connected using a serial to USB adapter, exposing `/dev/ttyUSB0` to dosemu2 as a serial port.

I'd have to dig up the driver, but I remember something like "4D", the hardware it was included with apparently also supported horizontal scrolling. It also included some extra files I think, I guess to send the right scroll events to the window as I suspect win31 didn't handle scroll input explicitly.

**Comment by javispedro** [comment]
As discussed in the vmwmouse thread, I have yet another mouse.drv (in Watcom C, not ASM) that uses int33 only, and also  provides some wheel support based on vmwmouse's "call USER.DLL from within the driver" idea + supports windows with only horizontal scrollbars like this "4D" driver. 

My intention is to pair this with a DOS int33 driver I'm making for VirtualBox, but it can also work under emulated int33 drivers.  
Here it is on dosemu2:

https://user-images.githubusercontent.com/123961/163694756-d6e66ca4-a61c-47f5-93c4-d32e490ed914.mp4

Actually it works better on dosemu2 than on VirtualBox/real HW right now :(  (win386 eats PS/2 wheel mouse packets for lunch).

It "just" requires two changes to the dosemu int33 driver .
Patch here https://gist.github.com/javispedro/e7db8b82047c494ab8dbe81367ce3875

1. Support enlarging the max coordinate range by setting a virtual coordinate range larger than the screen size. i.e. if the screen size is 640x200, but one sets a larger virtual window/range of 1000x1000, then the actual coordinate range becomes 1000x1000. 
My driver initializes window size to 0x7FFF x 0x7FFF, and expects to be given coordinates in 0.. 0x7FFF range. 
(why 0x7FFF instead of 0xFFFF ? I don't know if other mouse drivers treat this as a signed value).
No WinOS2 hook (int66h) is required. 
As far as I can see this is actually already supported by both Cutemouse, MS Mouse and DOSBox, because SimCity 2k kinda requires it (it sets a window of around 5000x5000 and expects coordinates in that range). 
However the patch needs a bit more thought, since it actually _breaks_ Simcity2k in DOSemu. 
It seems that DOSemu is already scaling the coordinates based on ... the speed ? So this results in double-scaling. 

2. When we call the user mouse proc, set bit 8 of the status word (currently unused) to indicate that this event comes from an absolute pointer.
I map this bit to SF_ABSOLUTE in the windows driver. This way when the int33 driver knows there is no absolute mouse (e.g. mouse is ~~not~~ grabbed)  it'll allow Windows to handle mouse input as relative including speed, acceleration, etc. as usual. 

Here is [source](https://git.javispedro.com/cgit/vbados.git/tree/mousew16.c) and [a zip with binary vbmouse.drv file](https://depot.javispedro.com/vbox/vbados/vbados.zip), if you want to try.

**Comment by NattyNarwhal** [comment]
I wonder if rebuilding VKD (source is in the DDK, after all) with a patch to make it not drop PS/2 wheel packets on the floor would help?

The annoying part is you couldn't really pack the VKD with the `.inf`, because for mouse drivers, it would only alter the VMD.

**Comment by stsp** [comment]
> (e.g. mouse is not grabbed)

"not grabbed" means absolute, so
this is inverted.

> it'll allow Windows to handle mouse input as relative including speed, acceleration, etc. as usual.

Convincing, applied.
Thanks!

> but one sets a larger virtual window/range of 1000x1000, then the actual coordinate range becomes 1000x1000.

Yes but the change you propose, will
likely actually affect speed, not window
size.
You need to run SerfCity game which
also sets the large window. If your change
doesn't break it in ungrabbed mouse mode,
then its interesting, but I think it will break
cursor speed.

But overall: you driver looks cool and
your video looks very cool too!
We just need to settle on the change
that you propose, and then our own
int33 windows driver can just go away.

**Comment by stsp** [comment]
> No WinOS2 hook (int66h) is required.

Is int66 ever documented?
In dosemu2 it is implemented based
on a quick disasm of a winos2 driver,
so its likely very incomplete.

**Comment by javispedro** [comment]
> I wonder if rebuilding VKD (source is in the DDK, after all) with a patch to make it not drop PS/2 wheel packets on the floor would help?
> 
> The annoying part is you couldn't really pack the VKD with the `.inf`, because for mouse drivers, it would only alter the VMD.

Yeah, I am reading the source code of VKD in the DDK...  lead me to this nice article : http://www.os2museum.com/wp/jumpy-ps2-mouse-in-enhanced-mode-windows-3-x/ . I am also checking (as I mention in a comment there) MOUSEVKD from Microsoft, as well as "GVKD" from KYE/Genius (according to Wikipedia they shipped a mouse with scroll wheel in the Win3.x era, but it seems to use a different protocol). 

There are two problems with the standard VKD:
1. It configures the PS/2 mouse the way it likes it, and the prevents everyone else from touching it, not even the System VM. This means I can't switch the mouse to wheel mode in the first place. But even if I could...
2. It hardcodes a packet size of 3 internally, but wheel mice generally use 4. This causes latency and desync problems. 

Both are fixable, so maybe I'll try to make "wheelvkd" that configures wheel mouse by itself and uses flexible packet size internally , but it won't look pretty... and definitely not going to rewrite VKD :)
(sorry for the offtopic :) )

**Comment by javispedro** [comment]
> Yes but the change you propose, will likely actually affect speed, not window size. You need to run SerfCity game which also sets the large window. If your change doesn't break it in ungrabbed mouse mode, then its interesting, but I think it will break cursor speed.

Right now even if I use unpatched dosemu there are slight desyncs with SC2k, i.e. eventually the host cursor appears some lines below the guest cursor when I enter/leave the window. With the patch it is much worse.  I don't understand the mechanism yet, and I am not entirely sure this speed thing is required since SC2k also sets a pretty large speed to compensate for the large coordinate size. But I'll try to compare the speed on real hardware, and try to find SerfCity too.

> Is int66 ever documented? In dosemu2 it is implemented based on a quick disasm of a winos2 driver, so its likely very incomplete.

Only found this: http://www.edm2.com/index.php/DDDR/2_-_Seamless_Windows_Support#Windows_Driver_Calls_to_VWIN_-_Accessed_via_INT_66h 

Anyway, fwiw, could also come up with another int33 extension if necessary, but want to avoid it unless there is really no other way.

**Comment by stsp** [comment]
> Right now even if I use unpatched dosemu there are slight desyncs with SC2k

Many games do their own scaling.
In that case its impossible to correctly
work with them in ungrabbed mode,
but I think (and hope) we do the best
guess. It would be good if you check
dosbox to see if it does better job with
ungrabbed mouse.

> Anyway, fwiw, could also come up with another int33 extension if necessary, but want to avoid it unless there is really
> no other way.

Of course there is:
https://github.com/dosemu2/win31-mouse-driver/blob/master/src/INT33H.ASM#L107
You can use fn 0x26 and scale
from that to your max coords by hands.

**Comment by javispedro** [comment]
> Of course there is: https://github.com/dosemu2/win31-mouse-driver/blob/master/src/INT33H.ASM#L107 You can use fn 0x26 and scale from that to your max coords by hands.

This returns 640x200 in MS Mouse (and cutemouse), which as you may imagine makes the mouse very jumpy, specially when the real res is 1024x768 :)

**Comment by stsp** [comment]
Would you like to use int66/ah=0x80
extension? Its used by winos2 driver.

**Comment by javispedro** [comment]
Actually I prefer to use the window size since that works with my int33 driver & DOSBox-X as is. Going to investigate the PS/2 wheel issue a bit more, but then I'll get back to this.

**Comment by stsp** [comment]
https://git.javispedro.com/cgit/vbados.git/tree/mousew16.c#n194
I don't think you are using window
size even now. You scale to 0xffff.
int66/ah=0x80 does the same thing.

**Comment by javispedro** [comment]
https://git.javispedro.com/cgit/vbados.git/tree/mousew16.c#n293 -- this is where I make the int33/07h and 08h calls to set the window to 0...0x7fff.
(without this call, "real" int33 drivers use 0..640x200 coords). 

The scaling in https://git.javispedro.com/cgit/vbados.git/tree/mousew16.c#n194 is just to bring that from 0x7fff to 0xffff 

(Whether I could set the window to 0..0xFFFF directly: I don't know, since window coordinates are supposedly signed integers. Cutemouse at least assumes that and their own test program uses a -40...600 window).

**Comment by stsp** [comment]
> Whether I could set the window to 0..0xFFFF directly: I don't know

This is exactly what int66/ah=0x80 does.

**Comment by javispedro** [comment]
I know, but it is another extension, e.g. DOSBox doesn't support it (nor would any other DOS int33 driver).

**Comment by stsp** [comment]
Probably you can then use 0x26
to see if the function worked?

**Comment by stsp** [comment]
Another idea is to amplify
sensitivity (fn 0x1a) after
enlarging a window?

**Comment by javispedro** [comment]
> Another idea is to amplify sensitivity (fn 0x1a) after enlarging a window?

I like that, will try.

**Comment by javispedro** [comment]
Anyway, this is what I mean that SC2k is already jumpy even with unpatched dosemu:

https://user-images.githubusercontent.com/123961/163823697-9577d403-d90b-4969-b8df-c774b7ef9307.mp4

Notice the vertical desync.

I don't have this problem at all with my int33 driver in VBox: 

https://user-images.githubusercontent.com/123961/163823721-cd0e724c-e2e4-4df4-a619-db38bfa76528.mp4

Where I am not using the speed to scale the absolute coordinates -- just the window size. 
(Need to try with SerfCity).

**Comment by stsp** [comment]
We have emumouse tool.
Try `emumouse y 8` before starting sc2000.

**Comment by stsp** [comment]
Btw, sc2000 seems to do exactly
that: enlarges virtual window and
amplifies speed.
I do recommend to use sensitivity
instead of speed, but its the same
trick as you can do. :)

**Comment by javispedro** [comment]
Argh, I have been doing sensitivity incorrectly in my driver. int33.lst says that [sensitivity is the same as speed](https://github.com/FDOS/mouse/blob/dd268521dc76d323d0597d1b21f58be5b82c6c25/int33.lst#L305), but even cutemouse itself [treats sensitivity differently](https://github.com/FDOS/mouse/blob/dd268521dc76d323d0597d1b21f58be5b82c6c25/ctmouse.asm#L3254)! 

But yeah, I see SimCity does set speed to [1,1] which given screen coords 640x480 means scaled to 5120x"1920", at the same time it sets a window size 5120x3840 , so this explains the emumouse 8 y trick, the game is likely wrong. But to me it just shows we should use window size only for scaling , not speed. 

Also dosemu uses 100 as the default sensitivity, but cutemouse uses 50 as default (and 100 is the max), and DOSBox copied the formula from cutemouse.  What a mess =)

**Comment by stsp** [comment]
No problem to sync sensitivity with
other sources, as its rarely used. So
if it helps you, lets do that.
I am not sure why do you propose to
completely ignore speed.

**Comment by javispedro** [comment]
Okey I see the point. serfcity/settlers is just setting a window of "5000x5000" , not changing the mouse speed at all, then proceeds to compute its own deltas of the mouse cursor position instead of using mickeys (wtf) . As a consequence of the windowsize-based scaling the mouse moves too fast in DOSBox, while in dosemu is it usable (but I still get a lot of desyncs!). 

There's another problem with speed/sensitivity: the coordinate system also scales with the screen resolution. Real hw int33 drivers don't know about SVGA and so they will still assume 640x200-based coordinates, but dosemu int33 knows about SVGA and so it will give you 800x600-based coordinates.
So again I cannot do a universal thing. 

Sorry. I guess I'll just do the int66, then use int33/0x26 to see if int66 worked, otherwise do the 0x7FFF window thing for the rest.

**Comment by stsp** [comment]
> but I still get a lot of desyncs!

But desyncs are unavoidable,
I do not even count them in.

> but dosemu int33 knows about SVGA and so it will give you 800x600-based coordinates.

Try `emumouse s 1` to disable VESA modes.
I think they were needed for PC-GEOS.

> So again I cannot do a universal thing.

But you can query the resolution with
0x26, so why is it a problem that dosemu2
knows about VESA modes? You can
query the resolution, so why to assume
any particular resolution instead?

**Comment by javispedro** [comment]
I can query the resolution even with the Windows API, but the problem is that no other int33 driver will do the same. 

My intention was to set speed to [1,1] and then a window to [5120,7680]. 
If the screen res is 640x480, this works everywhere. 
Dosemu scales coordinates to 640 * 8 x 480 * 16 = 5120x7680 
Dosbox, my driver, etc. scales coordinates using the window size = 5120x7680 again.

However on 800x600.
Dosemu now scales coordinates to 800 * 8 x 600 * 16 = 6400x9600
The rest still use the window size = 5120x7680

int33/26 result does not change to reflect this (in any of the drivers). Also dosemu returns bx=1 incorrectly on int33/26 (which means error / driver disabled according to int33.lst and cutemouse src), should be 0 to indicate driver enabled.

**Comment by stsp** [comment]
But you can query the resolution
via 0x26 to calculate the proper
speed values from that.
Don't set speed to 1:1, calculate
it from 2 resolutions.

**Comment by javispedro** [comment]
This logic works in all big 4 (Dosbox(-X), dosemu, Cutemouse, MS mouse):

```
// Set the speed to 1,1 to enlarge dosemu coordinate range by 8x16.
// In other absolute drivers, this doesn't change coordinates nor actual speed (which is inherited from host)
// In normal relative drivers, we'll use the raw mickeys anyways, so speed should also have no effect.
int33_set_mouse_speed(1, 1);

// Get what the int33 driver thinks the current coordinate range is
max_x = 0; max_y = 0;
int33_get_max_coordinates(&max_x, &max_y);
if (!max_x || !max_y) { max_x = 640; max_y = 200; } // For really bad drivers

// Multiply that by 8x16 to match dosemu coordinate range
// (and also in case the driver-provided coordinate range is too small)
max_x *= 8;
max_y *= 16;

// And use that to define the window size which is what most
// absolute drivers will use to define the coordinate range
int33_set_horizontal_window(0, max_x);
int33_set_vertical_window(0, max_y);
```

I couldn't use:
* Sensitivity, since for some reason it affected the number of mickeys as given by Cutemouse (so it makes relative mouse either too slow or too fast). So I'm limited to changing speed (speed a.k.a. mickeysPerSecond doesn't change the number of mickeys reported by DOS drivers). 
*  int66, since for some reason it crashes real hardware. Yes I checked to see if the interrupt vector was defined, and surprisingly it is (no idea why, but it pointed to something in windows).

**Comment by javispedro** [comment]
[Pushed](https://git.javispedro.com/cgit/vbados.git/commit/?id=513aec2aa20fdfcd862acc4189aa3226c876a051). Here's a bin if someone wants to try.

[vbmousedrv.zip](https://github.com/dosemu2/dosemu2/files/8507551/vbmousedrv.zip)

**Comment by stsp** [comment]
Good algo!
But the driver is not yet.
When I start win31, if the host mouse
is initially inside the dosemu window,
then the very first mouse move makes
the windows cursor jump to the up-left
corner, resulting in a permanent desync.
Moving mouse outside of dosemu
window and back, fixes it until the
win31 restart.

**Comment by stsp** [comment]
Fixed it myself. :)

**Comment by stsp** [comment]
@Baron-von-Riedesel @jschwartzenberg 
would you like to test this driver?
It looks pretty cool.
Wheel works!

**Comment by stsp** [comment]
I think https://depot.javispedro.com/vbox/vbados/
still doesn't have the latest version.
@javispedro Please ping when you
update the version, so we can update
the docs and remove our own driver
which seems no longer relevant.

Thank you!

**Comment by javispedro** [comment]
I have updated them to the last versions. But there are no new changes to win16 mouse since the binary from https://github.com/dosemu2/dosemu2/issues/1552#issuecomment-1101667362 
The only changes are for the DOS driver which is not required here. 

We also need to see if the win16 driver is not too crashy . Not sure that all these win16 API functions are reentrant/safe to call from int33 handler. If they aren't, I'll have to move the scroll wheel logic into a usermode .DLL .

**Comment by NattyNarwhal** [comment]
My guess is the functions we're calling seem mostly interrupt safe, because they don't do much:

* GetCursorPos probably reads globals only
* Reading window longs like class, style, parent, etc. should be safe
* EnumChildWindows might have some recordkeeping on the stack, but it's, again, only reading values
* WindowFromPoint has to look probably at USER state thoroughly, but again, reading

The only problem ones I could think of might be PostMessage (has to write messages somewhere). 

Also note a few of the DDK drivers do call USER functions. (Some is also just display drivers trying to dig for GDI functions, or stuff with pluggable components like KEYBOARD.) Some even outright call MessageBox and PostMessage, which I think would be pretty complex. (Printer drivers also do it, but they seem universally written in C and are loaded pretty late in the game, so I think they're a special case of not being a special case.)

**Comment by stsp** [comment]
Just out of curiosity:
have you tried shutting down
windows native cursor and
enabling the int33 (host) cursor
instead?
Would be interesting to see how
it looks like, and also dosemu2
follows the simple coords scheme
in that case, that is found in other
emulators. Complex scheme is
only active when cursor is invisible.

**Comment by javispedro** [comment]
I didn't think it was possible (cursor rendering is done by video driver on 3.x and that's quite a beast).
But maybe it is possible by calling HideCursor() a lot of times.
And probably could even use [GetCursor ](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getcursor) to fetch the shape and convert it to an int33 shape. 
And need to think of a way to check if the int33 driver can render the cursor (at the current video mode).
Sounds interesting.

**Comment by stsp** [comment]
Ah, as far as I can recall, we
do not support cursor shape in
an ungrabbed mouse mode.
But that's likely not a big problem.

**Comment by NattyNarwhal** [comment]
@javispedro Unrelated but since we're on the same thread, regarding:

> VMware shared folders support is interesting, but there is very little documentation that I can find, no sample code, and no open source implementation.

in your repo, I believe there's a VMware HGFS client in [https://sites.google.com/site/chitchatvmback/vmtools](this guy's source for various utilities), some of which work on DOS, though it's quite old. Unfortunately when I last tried it, it didn't work with modern Workstation for some reason.

**Comment by javispedro** [comment]
> I believe there's a VMware HGFS client in [https://sites.google.com/site/chitchatvmback/vmtools](this guy's source for various utilities), some of which work on DOS, though it's quite old. Unfortunately when I last tried it, it didn't work with modern Workstation for some reason.

Found another VMware client http://www.ibiblio.org/pub/micro/pc-stuff/freedos/files/repositories/1.3/pkg-html/vmsmount.html , this one from FreeDOS. It's also Watcom C and quite decently written. I really could have used it while writing mine. We really need a better catalog of DOS software :)

**Comment by stsp** [comment]
@javispedro I am getting certificate
expired error from this link:
https://depot.javispedro.com/vbox/vbados/vbados.zip

**Comment by javispedro** [comment]
> @javispedro I am getting certificate expired error from this link: https://depot.javispedro.com/vbox/vbados/vbados.zip

Should be fixed (first time certbot setup for that machine...).

**Comment by stsp** [comment]
Thank you.
It was for quite long btw.
I was initially hoping it would
fix itself. :)
